### DISCUSSION TITLE
[scaling]: Concurrent query execution

### DISCUSSION THREAD

**Discussion by amoghnjoshi** [duckdb/duckdb#8203] [Q&A]
**Use case:** I am using duckdb (python API) to periodically execute a fixed batch of queries (read-only connection). 
**Data**: Around 70 million rows with 20 columns
**Queries**: Involve filters, aggregations, distinct operator then group by and select a few rows
**Resources**: 16/32 CPUs

The objective is to finish the given batch of queries in minimum time, making the best use of CPUs available.

**Observation**:
The simple/default approach is to use the intra-query parallelism offered by duckdb. 
Create just one connection and configure it use all the cores with the _threads, worker_threads_ configuration.
However, the observation is that the performance improves with the _threads_ configuration as we go from threads=1 to threads=4 or even 8. After this the improvement is not that significant. (tested with table having 12 million rows)
Probably, these additional CPUs can be better utilised to execute other queries concurrently (inter-query parallelism).

Under this condition, what could be the recommended ways for concurrent query execution to efficiently use all the CPUs available? The aim is to finish the whole workload as quickly as possible. Individual queries taking longer time is okay if the total time is minimised.
1. Create multiple duplicate (**cursor**) connections, derived from a parent connection which is having **threads=#total_cores** Or
2. Create multiple (N) **separate** connections, with each connection using **threads=(#total_cores/N)** 
In either case, create multiple application threads, using the individual connection to pump queries.

Any other alternative is welcome.
Also, is there any role of the configuration _external_threads_ in the case?

**Comment by l1t1** [comment]
if some queries can share the same temporary result, you can save them first, like MV in some db.

**Reply by amoghnjoshi** [reply]
Thanks @l1t1 , such custom handling would be the last resort.

What could be the recommended way for concurrent query execution from #1 and #2 above?

**Comment by amoghnjoshi** [comment]
@hannes , can you please make any comment on this?
